File: capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
  int sz;
  bool func(const int&c,const vector<int>&v,int d){
    int cap=c;
      for(int k=0;k<sz;++k){
        if(c==11)
            cout<<cap<<' '<<v[k]<<' '<<d<<'\n';
        if(cap<v[k]){
          if(d<2)
            return false;
          --d;
          cap=c-v[k];
        }
        else
          cap-=v[k];
      }
    return true;
  }
    int shipWithinDays(vector<int>&w,int&d) {
      sz=w.size();
      int res,mx=0;
      mx=accumulate(w.begin(),w.end(),mx);
      res=mx;
      int mn=*max_element(w.begin(),w.end());
      //if(d==1)
        //return res;
      while(mn<=mx){
        int md=mn+((mx-mn)>>1);
        if(func(md,w,d)){
          res=min(res,md);
          mx=md-1;
        }
      else{
        //cout<<"false for "<<md<<'\n';
        mn=md+1;
      }
    }
    return res;
  }
};

File: capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.java
    class Solution {
        private Integer sz;

        boolean func(int[] w, Integer d, Integer m) {
            Integer dReq = 1, cap = 0;
            for (Integer i=0; i < sz; ++i) {
                cap += w[i];
                if (cap > m) {
                    ++dReq;
                    if(dReq>d)
                        return false;
                    cap = w[i];
                }
            }
            return true;
        }

        public int shipWithinDays(int[] w, int d) {
            sz = w.length;
            Integer l = w[0], h = 0;
            // System.out.println(res);
            for (Integer i : w) {
                l = Math.max(l, i);
                h += i;
            }
            Integer res = h;
            while (l <= h) {
                Integer m = l + ((h - l) >> 1);
                if (func(w, d, m)) {
                    res = Math.min(res, m);
                    h = m - 1;
                } else
                    l = m + 1;
            }
            return res;
        }
    }